File: code/Java/syntatic/SyntaticAnalysis.java
package syntatic;

import interpreter.command.AssignCommand;
import interpreter.command.BlocksCommand;
import interpreter.command.Command;
import interpreter.command.IfCommand;
import interpreter.command.OutputCommand;
import interpreter.command.WhileCommand;
import interpreter.expr.BinaryIntExpr;
import interpreter.expr.BoolExpr;
import interpreter.expr.ConstBoolExpr;
import interpreter.expr.ConstIntExpr;
import interpreter.expr.IntExpr;
import interpreter.expr.IntOp;
import interpreter.expr.NegIntExpr;
import interpreter.expr.NotBoolExpr;
import interpreter.expr.ReadIntExpr;
import interpreter.expr.RelOp;
import interpreter.expr.SingleBoolExpr;
import interpreter.expr.Variable;
import lexical.Lexeme;
import lexical.LexicalAnalysis;
import lexical.TokenType;

public class SyntaticAnalysis {

    private LexicalAnalysis lex;
    private Lexeme current;

    public SyntaticAnalysis(LexicalAnalysis lex) {
        this.lex = lex;
        this.current = lex.nextToken();
    }

    private void advance() {
        current = lex.nextToken();
    }

    private void eat(TokenType type) {
        // System.out.println("Expected (..., " + type + "), found (\"" +
        //     current.token + "\", " + current.type + ")");
        if (type == current.type) {
            advance();
        } else {
            showError();
        }
    }

    private void showError() {
        System.out.printf("%02d: ", lex.getLine());

        switch (current.type) {
            case INVALID_TOKEN:
                System.out.printf("Lexema inválido [%s]\n", current.token);
                break;
            case UNEXPECTED_EOF:
            case END_OF_FILE:
                System.out.printf("Fim de arquivo inesperado\n");
                break;
            default:
                System.out.printf("Lexema não esperado [%s]\n", current.token);
                break;
        }

        System.exit(1);
    }

    public BlocksCommand start() {
        BlocksCommand cmds = procProgram();
        eat(TokenType.END_OF_FILE);

        return cmds;
    }

    // <program>   ::= program <cmdlist>
    private BlocksCommand procProgram() {
        eat(TokenType.PROGRAM);
        return procCmdList();
    }

    // <cmdlist>   ::= <cmd> { <cmd> }
    private BlocksCommand procCmdList() {
        // There is no precise way to track the line number
        // for block commands, thus using -1.
        BlocksCommand cb = new BlocksCommand(-1);

        Command c = procCmd();
        cb.addCommand(c);
        while (current.type == TokenType.VAR ||
                current.type == TokenType.OUTPUT ||
                current.type == TokenType.IF ||
                current.type == TokenType.WHILE) {
            c = procCmd();
            cb.addCommand(c);
        }

        return cb;
    }

    // <cmd>       ::= (<assign> | <output> | <if> | <while>) ;
    private Command procCmd() {
        Command c = null;
        if (current.type == TokenType.VAR)
            c = procAssign();
        else if (current.type == TokenType.OUTPUT)
            c = procOutput();
        else if (current.type == TokenType.IF)
            c = procIf();
        else if (current.type == TokenType.WHILE)
            c = procWhile();
        else
            showError();

        eat(TokenType.SEMICOLON);

        return c;
    }

    // <assign>    ::= <var> = <intexpr>
    private AssignCommand procAssign() {
        Variable var = procVar();
        int line = lex.getLine();

        eat(TokenType.ASSIGN);
        IntExpr expr = procIntExpr();

        return new AssignCommand(line, var, expr);
    }

    // <output>    ::= output <intexpr>
    private OutputCommand procOutput() {
        eat(TokenType.OUTPUT);
        int line = lex.getLine();

        IntExpr expr = procIntExpr();

        return new OutputCommand(line, expr);
    }

    // <if>        ::= if <boolexpr> then <cmdlist> [ else <cmdlist> ] done
    private IfCommand procIf() {
        eat(TokenType.IF);
        int line = lex.getLine();

        BoolExpr cond = procBoolExpr();
        eat(TokenType.THEN);
        BlocksCommand thenCmds = procCmdList();

        BlocksCommand elseCmds = null;
        if (current.type == TokenType.ELSE) {
            advance();
            elseCmds = procCmdList();
        }
        eat(TokenType.DONE);

        return new IfCommand(line, cond, thenCmds, elseCmds);
    }

    // <while>     ::= while <boolexpr> do <cmdlist> done
    private WhileCommand procWhile() {
        eat(TokenType.WHILE);
        int line = lex.getLine();

        BoolExpr cond = procBoolExpr();
        eat(TokenType.DO);
        Command cmds = procCmdList();
        eat(TokenType.DONE);

        return new WhileCommand(line, cond, cmds);
    }

    // <boolexpr>  ::= false | true |
    //                 not <boolexpr> |
    //                 <intterm> (== | != | < | <= | > | >=) <intterm>
    private BoolExpr procBoolExpr() {
        if (current.type == TokenType.FALSE) {
            advance();
            return new ConstBoolExpr(lex.getLine(), false);
        } else if (current.type == TokenType.TRUE) {
            advance();
            return new ConstBoolExpr(lex.getLine(), true);
        } else if (current.type == TokenType.NOT) {
            advance();
            int line = lex.getLine();
            BoolExpr tmp = procBoolExpr();
            return new NotBoolExpr(line, tmp);
        } else {
            IntExpr left = procIntTerm();
            int line = lex.getLine();

            RelOp op  = null;
            switch (current.type) {
                case EQUAL:
                    op = RelOp.Equal;
                    break;
                case NOT_EQUAL:
                    op = RelOp.NotEqual;
                    break;
                case LOWER:
                    op = RelOp.LowerThan;
                    break;
                case LOWER_EQUAL:
                    op = RelOp.LowerEqual;
                    break;
                case GREATER:
                    op = RelOp.GreaterThan;
                    break;
                case GREATER_EQUAL:
                    op = RelOp.GreaterEqual;
                    break;
                default:
                    showError();
                    break;
            }

            advance();

            IntExpr right = procIntTerm();

            return new SingleBoolExpr(line, left, op, right);
        }
    }

    // <intexpr>   ::= [ + | - ] <intterm> [ (+ | - | * | / | %) <intterm> ]
    private IntExpr procIntExpr() {
        boolean isNegative = false;
        if (current.type == TokenType.ADD) {
            advance();
        } else if (current.type == TokenType.SUB) {
            advance();
            isNegative = true;
        }

        IntExpr left;
        if (isNegative) {
            int line = lex.getLine();
            IntExpr tmp = procIntTerm();
            left = new NegIntExpr(line, tmp);
        } else {
            left = procIntTerm();
        }

        if (current.type == TokenType.ADD |
                current.type == TokenType.SUB |
                current.type == TokenType.MUL |
                current.type == TokenType.DIV |
                current.type == TokenType.MOD) {
            int line = lex.getLine();

            IntOp op;
            switch (current.type) {
                case ADD:
                    op = IntOp.Add;
                    break;
                case SUB:
                    op = IntOp.Sub;
                    break;
                case MUL:
                    op = IntOp.Mul;
                    break;
                case DIV:
                    op = IntOp.Div;
                    break;
                case MOD:
                default:
                    op = IntOp.Mod;
                    break;
            }

            advance();

            IntExpr right = procIntTerm();

            return new BinaryIntExpr(line, left, op, right);
        } else {
            return left;
        }
    }

    // <intterm>   ::= <var> | <const> | read
    private IntExpr procIntTerm() {
        if (current.type == TokenType.VAR) {
            return procVar();
        } else if (current.type == TokenType.NUMBER) {
            return procConst();
        } else {
            eat(TokenType.READ);
            int line = lex.getLine();
            return new ReadIntExpr(line);
        }
    }

    // <var>       ::= id
    private Variable procVar() {
        String name = current.token;
        eat(TokenType.VAR);

        return Variable.instance(name);
    }

    // <const>     ::= number
    private ConstIntExpr procConst() {
        String tmp = current.token;
        eat(TokenType.NUMBER);
        int line = lex.getLine();

        int number;
        try {
            number = Integer.parseInt(tmp);
        } catch (Exception e) {
            number = 0;
        }

        return new ConstIntExpr(line, number);
    }

}

File: code/C++/syntatic/SyntaticAnalysis.cpp
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstdlib>

#include "SyntaticAnalysis.h"

#include "../interpreter/expr/IntExpr.h"
#include "../interpreter/expr/BinaryIntExpr.h"
#include "../interpreter/expr/ConstIntExpr.h"
#include "../interpreter/expr/NegIntExpr.h"
#include "../interpreter/expr/ReadIntExpr.h"
#include "../interpreter/expr/Variable.h"
#include "../interpreter/expr/BoolExpr.h"
#include "../interpreter/expr/ConstBoolExpr.h"
#include "../interpreter/expr/NotBoolExpr.h"
#include "../interpreter/expr/SingleBoolExpr.h"
#include "../interpreter/command/Command.h"
#include "../interpreter/command/AssignCommand.h"
#include "../interpreter/command/BlocksCommand.h"
#include "../interpreter/command/IfCommand.h"
#include "../interpreter/command/OutputCommand.h"
#include "../interpreter/command/WhileCommand.h"

SyntaticAnalysis::SyntaticAnalysis(LexicalAnalysis& lex) :
	m_lex(lex), m_current(lex.nextToken()) {
}

SyntaticAnalysis::~SyntaticAnalysis() {
}

Command* SyntaticAnalysis::start() {
	Command* cmd = procProgram();
	eat(TT_END_OF_FILE);
	return cmd;
}

void SyntaticAnalysis::advance() {
	m_current = m_lex.nextToken();
}

void SyntaticAnalysis::eat(enum TokenType type) {
	// std::cout << "Expected (..., " << tt2str(type)
	//           <<  "), found (\"" + m_current.token << "\", "
	// 		  << tt2str(m_current.type) << ")" << std::endl;
	if (type == m_current.type) {
		advance();
	} else {
		showError();
	}
}

void SyntaticAnalysis::showError() {
	std::cout << std::setw(2) << std::setfill('0') << m_lex.line() << ": ";

	switch (m_current.type) {
		case TT_INVALID_TOKEN:
			std::cout << "Lexema inválido [" << m_current.token << "]" << std::endl;
			break;
		case TT_UNEXPECTED_EOF:
		case TT_END_OF_FILE:
			std::cout << "Fim de arquivo inesperado" << std::endl;
			break;
		default:
			std::cout << "Lexema não esperado [" << m_current.token << "]" << std::endl;
			break;
	}

    exit(1);
}

// <program>   ::= program <cmdlist>
Command* SyntaticAnalysis::procProgram() {
	eat(TT_PROGRAM);
	Command* cmd = procCmdList();
	return cmd;
}

// <cmdlist>   ::= <cmd> { <cmd> }
BlocksCommand* SyntaticAnalysis::procCmdList() {
	int line = m_lex.line();
	BlocksCommand* cmds = new BlocksCommand(line);

	Command* cmd = procCmd();
	cmds->addCommand(cmd);

	while (m_current.type == TT_VAR ||
	       m_current.type == TT_OUTPUT ||
		   m_current.type == TT_IF ||
		   m_current.type == TT_WHILE) {
		cmd = procCmd();
		cmds->addCommand(cmd);
	}

	return cmds;
}

// <cmd>       ::= (<assign> | <output> | <if> | <while>) ;
Command* SyntaticAnalysis::procCmd() {
	Command* cmd = 0;
	if (m_current.type == TT_VAR) {
		cmd = procAssign();
	} else if (m_current.type == TT_OUTPUT) {
		cmd = procOutput();
	} else if (m_current.type == TT_IF) {
		cmd = procIf();
	} else if (m_current.type == TT_WHILE) {
		cmd = procWhile();
	} else {
		showError();
	}

	eat(TT_SEMICOLON);
	return cmd;
}

// <assign>    ::= <var> = <intexpr>
AssignCommand* SyntaticAnalysis::procAssign() {
	int line = m_lex.line();

	Variable* var = procVar();
	eat(TT_ASSIGN);
	IntExpr* expr = procIntExpr();

	AssignCommand* cmd = new AssignCommand(line, var, expr);
	return cmd;
}

// <output>    ::= output <intexpr>
OutputCommand* SyntaticAnalysis::procOutput() {
	eat(TT_OUTPUT);
	int line = m_lex.line();

	IntExpr* expr = procIntExpr();
	OutputCommand* cmd = new OutputCommand(line, expr);
	return cmd;
}

// <if>        ::= if <boolexpr> then <cmdlist> [ else <cmdlist> ] done
IfCommand* SyntaticAnalysis::procIf() {
	eat(TT_IF);
	int line = m_lex.line();

	BoolExpr* cond = procBoolExpr();
	eat(TT_THEN);
	Command* thenCmds = procCmdList();
	Command* elseCmds = 0;
	if (m_current.type == TT_ELSE) {
		advance();
		elseCmds = procCmdList();
	}
	eat(TT_DONE);

	IfCommand* cmd = new IfCommand(line, cond, thenCmds, elseCmds);
	return cmd;
}

// <while>     ::= while <boolexpr> do <cmdlist> done
WhileCommand* SyntaticAnalysis::procWhile() {
	eat(TT_WHILE);
	int line = m_lex.line();

	BoolExpr* expr = procBoolExpr();
	eat(TT_DO);
	Command* cmds = procCmdList();
	eat(TT_DONE);

	WhileCommand* cmd = new WhileCommand(line, expr, cmds);
	return cmd;
}

// <boolexpr>  ::= false | true |
//                 not <boolexpr> |
//                 <intterm> (== | != | < | > | <= | >=) <intterm>
BoolExpr* SyntaticAnalysis::procBoolExpr() {
	if (m_current.type == TT_FALSE) {
		advance();
		return new ConstBoolExpr(m_lex.line(), false);
	} else if (m_current.type == TT_TRUE) {
		advance();
		return new ConstBoolExpr(m_lex.line(), false);
	} else if (m_current.type == TT_NOT) {
		advance();
		int line = m_lex.line();
		BoolExpr* expr = procBoolExpr();
		return new NotBoolExpr(line, expr);
	} else {
		int line = m_lex.line();
		IntExpr* left = procIntTerm();

		enum SingleBoolExpr::Op op = SingleBoolExpr::EQUAL;
		switch (m_current.type) {
			case TT_EQUAL:
				op = SingleBoolExpr::EQUAL;
				advance();
				break;
			case TT_NOT_EQUAL:
				op = SingleBoolExpr::NOT_EQUAL;
				advance();
				break;
			case TT_LOWER:
				op = SingleBoolExpr::LOWER;
				advance();
				break;
			case TT_GREATER:
				op = SingleBoolExpr::GREATER;
				advance();
				break;
			case TT_LOWER_EQUAL:
				op = SingleBoolExpr::LOWER_EQUAL;
				advance();
				break;
			case TT_GREATER_EQUAL:
				op = SingleBoolExpr::GREATER_EQUAL;
				advance();
				break;
			default:
				showError();
		}

		IntExpr* right = procIntTerm();

		BoolExpr* expr = new SingleBoolExpr(line, left, op, right);
		return expr;
	}
}

// <intexpr>   ::= [ + | - ] <intterm> [ (+ | - | * | / | %) <intterm> ]
IntExpr* SyntaticAnalysis::procIntExpr() {
	bool is_negative = false;
	if (m_current.type == TT_ADD) {
		advance();
	} else if (m_current.type == TT_SUB) {
		advance();
		is_negative = true;
	}

	IntExpr* left;
	if (is_negative) {
		int line = m_lex.line();
		IntExpr* tmp = procIntTerm();
		left = new NegIntExpr(line, tmp);
	} else {
		left = procIntTerm();
	}

	if (m_current.type == TT_ADD ||
	    m_current.type == TT_SUB ||
		m_current.type == TT_MUL ||
	    m_current.type == TT_DIV ||
	    m_current.type == TT_MOD) {
		int line = m_lex.line();

		enum BinaryIntExpr::Op op;
		switch (m_current.type) {
			case TT_ADD:
				op = BinaryIntExpr::ADD;
				advance();
				break;
			case TT_SUB:
				op = BinaryIntExpr::SUB;
				advance();
				break;
			case TT_MUL:
				op = BinaryIntExpr::MUL;
				advance();
				break;
			case TT_DIV:
				op = BinaryIntExpr::DIV;
				advance();
				break;
			case TT_MOD:
			default:
				op = BinaryIntExpr::MOD;
				advance();
				break;
		}

		IntExpr* right = procIntTerm();

		left = new BinaryIntExpr(line, left, op, right);
	}

	return left;
}

// <intterm>   ::= <var> | <const> | read
IntExpr* SyntaticAnalysis::procIntTerm() {
	if (m_current.type == TT_VAR) {
		return procVar();
	} else if (m_current.type == TT_NUMBER) {
		return procConst();
	} else {
		eat(TT_READ);
		int line = m_lex.line();
		ReadIntExpr* expr = new ReadIntExpr(line);
		return expr;
	}
}

// <var>       ::= id
Variable* SyntaticAnalysis::procVar() {
	std::string tmp = m_current.token;

	eat(TT_VAR);
	int line = m_lex.line();

	Variable* var = new Variable(line, tmp);
	return var;
}

// <const>     ::= number
ConstIntExpr* SyntaticAnalysis::procConst() {
	std::string tmp = m_current.token;

	eat(TT_NUMBER);
	int line = m_lex.line();

	int value = atoi(tmp.c_str());
	ConstIntExpr* expr = new ConstIntExpr(line, value);
	return expr;
}
